### PR TITLE
hugging Face Datasets Plugin

### PR DESCRIPTION
**Pull request by esadler-hbo**
# TL;DR
Hugging Face provides great packages to make working with state-of-the-art language models easy. Integrating with Flyte would connect ETL to the training and inference of deep learning models seamlessly. 

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
You can use Hugging face to create high quality embeddings, which is becoming really valuable to a lot of companies. Flyte could elegantly handle the different infrastructure considerations. Notice there is no model training, which makes this workflow especially great. 

The first integration is adding Hugging Face's [datasets](https://github.com/huggingface/datasets) into Flyte's `StructuredDatasets`. Their datasets is a very performant way to pass data into neural networks. It is based on `tf.data.Dataset`, but uses Arrow instead of TFRecords. I am excited by the idea of having an ETL job output a `pyspark.sql.DataFrame` and then doing batch training and batch inference with a Hugging Face dataset seamlessly.

The second integration would be coming up highly scalable task for step 2 in the following workflow: 
1. ETL: prepare dataset of text
2. Inference: run data through a Hugging Face model pipeline
3. Upload: Push results to a database that can handle vectors, like [Pinecone](https://www.pinecone.io/)

I have heard from @gdj0nes that this is common workflow that has infra pain points.

Finally, Hugging Face has a platform where you can save datasets, models, and deploy ML applications. There are opportunities to integrate with their platform that should be mentioned, but are lower priority.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by kumare3** [comment]
@esadler-hbo you rock! Love all 3 integration goals.

@esadler-hbo & @samhita-alla would you folks be open to writing a blog?

File: plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    def encode(
+        self,
+        ctx: FlyteContext,
+        structured_dataset: StructuredDataset,
+        structured_dataset_type: StructuredDatasetType,
+    ) -> literals.StructuredDataset:
+        df = typing.cast(datasets.Dataset, structured_dataset.dataframe)
+
+        local_dir = ctx.file_access.get_random_local_directory()
+        local_path = f"{local_dir}/00000"
```
Comment (kumare3) [inline plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py:34]: @wild-endeavor / @pingsutw is it possible to directly stream write instead of write a file and upload directory. I thought we had done this?

File: plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        df.to_parquet(local_path)
+
+        remote_dir = typing.cast(str, structured_dataset.uri) or ctx.file_access.get_random_remote_directory()
+        ctx.file_access.upload_directory(local_dir, remote_dir)
+        return literals.StructuredDataset(uri=remote_dir, metadata=StructuredDatasetMetadata(structured_dataset_type))
+
+
+class ParquetToHuggingFaceDatasetDecodingHandler(StructuredDatasetDecoder):
+    def __init__(self, protocol: str):
+        super().__init__(datasets.Dataset, protocol, PARQUET)
```
Comment (kumare3) [inline plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py:45]: We do not need the protocol - cc @wild-endeavor / @pingsutw ? Why? This should be decoupled from the persistence layer

File: flytekit/types/structured/structured_dataset.py [hunk 3]
```diff
@@ -668,6 +671,8 @@ def to_html(self, ctx: FlyteContext, python_val: typing.Any, expected_python_typ
         elif importlib.util.find_spec("polars") is not None and isinstance(df, pl.DataFrame):
             describe_df = df.describe()
             return pd.DataFrame(describe_df.transpose(), columns=describe_df.columns).to_html(index=False)
+        elif importlib.util.find_spec("datasets") is not None and isinstance(df, datasets.arrow_dataset.Dataset):
```
Comment (kumare3) [inline flytekit/types/structured/structured_dataset.py:674]: Cc @eapolinario / @pingsutw why is this not overloaded by the plugins?

**Comment by wild-endeavor** [comment]
@esadler-hbo let me take a look at this and try to fix some of the handling around protocol

**Comment by esadler-hbo** [comment]
@wild-endeavor amazing! I’ll get a chance to work on this more this weekend.

File: flytekit/types/structured/structured_dataset.py (hunk 3 above)
Comment (pingsutw) [reply flytekit/types/structured/structured_dataset.py:674]: We only register the decoder and encoder in the plugins. yes, the better approach is to implement `to_html()` in the flytekit-plugins, and auto loads it when the plugin is installed.

https://github.com/flyteorg/flytekit/blob/31277351c831b195951f1bc1152a4d08338e5633/plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py#L22-L63

**Comment by wild-endeavor** [comment]
Thanks!  Yeah I really need to get to those changes I was talking about today.

**Comment by esadler-hbo** [comment]
@kumare3 @wild-endeavor I am going to get back to this soon. I think downscoping to just the datasets python package is probably the right way to go.

It would be useful for unionml and my workflows at work when I want to use pytorch after a spark preprocessing job.

File: flytekit/types/structured/structured_dataset.py (hunk 3 above)
Comment (peridotml) [reply flytekit/types/structured/structured_dataset.py:674]: @pingsutw I tried to look for an alternative way of adding `to_html` like on the custom types, but I couldn't figure out what to do. Do you have any examples on how I should do this?

File: plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py (hunk 1 above)
Comment (kumare3) [reply plugins/flytekit-huggingface/flytekitplugins/huggingface/sd_transformers.py:34]: Ping

**Comment by wild-endeavor** [comment]
tagged you also on the other PR @easadler-hbo

File: plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import datasets
+import flytekitplugins.huggingface  # noqa F401
+import pandas as pd
+
+try:
+    from typing import Annotated
+except ImportError:
+    from typing_extensions import Annotated
```
Comment (pingsutw) [inline plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py:8]:
```suggestion
from typing_extensions import Annotated
```

File: plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from flytekit.types.structured.structured_dataset import PARQUET, StructuredDataset
+
+subset_schema = Annotated[StructuredDataset, kwtypes(col2=str), PARQUET]
+full_schema = Annotated[StructuredDataset, PARQUET]
+
+
+def test_huggingface_dataset_workflow_subset():
+    @task
+    def generate() -> subset_schema:
+        df = pd.DataFrame({"col1": [1, 3, 2], "col2": list("abc")})
```
Comment (pingsutw) [inline plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py:20]: Could we use `datasets.Dataset` here? we have to test that we can successfully convert the dataset to parquet.

File: plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        return consume(df=generate())
+
+    result = wf()
+    assert result is not None
+
+
+def test_huggingface_dataset__workflow_full():
+    @task
+    def generate() -> full_schema:
+        df = pd.DataFrame({"col1": [1, 3, 2], "col2": list("abc")})
```
Comment (pingsutw) [inline plugins/flytekit-huggingface/tests/test_huggingface_plugin_sd.py:44]: same here

File: plugins/flytekit-huggingface/flytekitplugins/huggingface/__init__.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+"""
+.. currentmodule:: flytekitplugins.polars
```
Comment (pingsutw) [inline plugins/flytekit-huggingface/flytekitplugins/huggingface/__init__.py:2]:
```suggestion
.. currentmodule:: flytekitplugins.huggingface
```

**Review by pingsutw — COMMENTED** [review]
@esadler-hbo Thanks, Overall LGTM. Some minor comments below.

**Review by pingsutw — APPROVED** [review]
